### PR TITLE
Feat: Add a cron job to hit the filter orphan tasks API every 6hours

### PR DESCRIPTION
**Pull request by MehulKChaudhari**
Date: 2/04/2024

Developer Name: @MehulKChaudhari 

---

## Issue Ticket Number

- https://github.com/Real-Dev-Squad/todo-action-items/issues/215
- https://github.com/Real-Dev-Squad/website-backend/pull/1996

## Description

This PR add's cron job which run every 6hours and hits orphan tasks API, That API filters all the tasks from users who are not in discord server and the tasks which are not `DONE` are marked as `BACKLOG`

##Warning ⚠️ 

Please merge this PR after https://github.com/Real-Dev-Squad/website-backend/pull/1996

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [x] Yes
- [ ] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/handlers/scheduledEventHandler.ts [hunk 1]
```diff
@@ -141,3 +147,52 @@ export const syncIdle7dUsers = async (env: env) => {
...
+	const namespace = env[NAMESPACE_NAME] as unknown as KVNamespace;
+	let lastOrphanTasksFilteration: string | null = '0';
+	try {
+		lastOrphanTasksFilteration = await namespace.get('ORPHAN_TASKS_UPDATED_TIME');
+		if (lastOrphanTasksFilteration === null) {
+			console.error('Error while fetching KV "ORPHAN_TASKS_UPDATED_TIME" timestamp');
+		}
+		if (!lastOrphanTasksFilteration) {
+			lastOrphanTasksFilteration = '0';
+		}
```
Comment (yesyash) [inline src/handlers/scheduledEventHandler.ts]: - as we're using the `!` operator console.error can be moved to the if statement below.
- if we're setting the value of `lastOrphanTasksFilteration` in the last if statement why are we logging an error above?
- for instances where the value is set for the first time for the namespace the value returned from `namespace.get` will be null in which case creating an error log does not seem to be the right thing to do, what do you think?

**Review by yesyash — CHANGES_REQUESTED** [review]

File: src/handlers/scheduledEventHandler.ts (hunk 1 above)
Comment (MehulKChaudhari) [reply src/handlers/scheduledEventHandler.ts]: 1st point: Makes sense making the change.
2nd Point: if there is some issue while getting the `lastOrphanTasksFilteration` but makes sense to put it in second if statement
3rd point: Yes right but what in the cases when we have the timestamp saved but something went wrong?

**Review by Achintya-Chatterjee — APPROVED** [review]

File: src/handlers/scheduledEventHandler.ts (hunk 1 above)
Comment (MehulKChaudhari) [reply src/handlers/scheduledEventHandler.ts]: Made the changes.

Comment (yesyash) [reply src/handlers/scheduledEventHandler.ts]: >2nd Point: if there is some issue while getting the lastOrphanTasksFilteration but makes sense to put it in second if statement
3rd point: Yes right but what in the cases when we have the timestamp saved but something went wrong?

we already have a catch block for it if something goes wrong while fetching data the catch block will be triggered.

**Review by yesyash — APPROVED** [review]

**Review by lakshayman — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]

File: src/worker.ts [hunk 2]
```diff
@@ -20,7 +21,10 @@ export default {
 	async scheduled(req: ScheduledController, env: env, ctx: ExecutionContext) {
 		switch (req.cron) {
 			case EVERY_6_HOURS: {
-				return await callDiscordNicknameBatchUpdate(env);
+				await callDiscordNicknameBatchUpdate(env);
+				await filterOrphanTasks(env);
+				console.log('Worker for filtering the orphan tasks has completed');
```
Comment (prakashchoudhary07) [inline src/worker.ts:26]: await? won't this cause workers to exhaust 30s timelimit

Comment (MehulKChaudhari) [reply src/worker.ts:26]: Sir, without `await` how we can call the async function, I mean yes you are right the case is real where we might be exhausting the 30 time-limit. But I then how to do it without `await`

**Pull request merged**